clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubBasedAdminClientManager.java

```diff
@@ -114,96 +109,90 @@ private Future<SubscriptionAdminClient> getOrCreateSubscriptionAdminClient() {
+        return vertx.executeBlocking(promise -> promise.tryComplete(client.getTopic(topic).getName()));
+    }
 
+    private Future<String> createTopic(final TopicName topic, final TopicAdminClient client) {
+        final Future<String> createdTopic = vertx
+                .executeBlocking(promise -> promise.tryComplete(client.createTopic(topic).getName()));
+        createdTopic.onSuccess(top -> LOG.debug("Topic {} created successfully.", topic))
+                .onFailure(thr -> LOG.debug("Creating topic failed [topic: {}, projectId: {}]", topic, projectId));
+        return createdTopic;
     }
 
     /**
-     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and prefix.
+     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and
+     * prefix.
      *
      * @param endpoint The endpoint name of the subscription, e.g. command_internal.
      * @param prefix The prefix of the subscription, e.g. the adapter instance ID.
-     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed
-     *         Future if it could not be created.
+     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed Future if
+     *         it could not be created.
      */
     public Future<String> getOrCreateSubscription(final String endpoint, final String prefix) {
         final String topicAndSubscriptionName = PubSubMessageHelper.getTopicName(endpoint, prefix);
         final TopicName topic = TopicName.of(projectId, topicAndSubscriptionName);
         final SubscriptionName subscription = SubscriptionName.of(projectId, topicAndSubscriptionName);
 
-        if (subscriptions.contains(subscription.toString())) {
-            LOG.debug("Subscription {} already exists, continue", subscription);
-            return Future.succeededFuture(subscription.getSubscription());
-        }
         return getOrCreateSubscriptionAdminClient()
                 .onFailure(thr -> LOG.debug("admin client creation failed", thr))
-                .compose(client -> getOrCreateSubscription(projectId, subscription, topic, client));
+                .compose(client -> getSubscription(subscription, client)
+                        .recover(thr -> {
+                            if (thr instanceof NotFoundException) {
+                                return createSubscription(subscription, topic, client);
+                            } else {
+                                return Future.failedFuture(thr);
+                            }
+                        }));
+    }
+
+    private Future<String> getSubscription(final SubscriptionName subscription, final SubscriptionAdminClient client) {
+        return vertx.executeBlocking(promise -> promise.tryComplete(client.getSubscription(subscription).getName()));
```

Comment:
Using `promise.tryComplete` here means that the promise should be completed with the value if it *has not yet been completed*. I suspect, however, that you intended to complete the promise only in case of the `client.getSubscription` invocation having succeeded, right?

If so, then you need to wrap the `client.getSubscription` invocation in a try/catch block and either invoke `promise.complete` or `promise.fail` based on the outcome ...

Reply:
You are right. Somehow it seemed to work like I had it before, but I definitely like your suggestion more so I implemented it.